reaper_test.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2024 The go-zeromq Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package zmq4
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"runtime"
+	"testing"
+	"time"
+)
+
+func TestConnReaperDeadlock2(t *testing.T) {
+	ep := must(EndPoint("tcp"))
+	defer cleanUp(ep)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Bind the server.
+	srv := NewRouter(ctx, WithLogger(Devnull)).(*routerSocket)
+	if err := srv.Listen(ep); err != nil {
+		t.Fatalf("could not listen on %q: %+v", ep, err)
+	}
+	defer srv.Close()
+
+	// Add modified clients connection to server
+	// so any send to client will trigger context switch
+	// and be failing.
+	// Idea is that while srv.Send is progressing,
+	// the connection will be closed and assigned
+	// for connection reaper, and reaper will try to remove those
+	id := "client-x"
+	srv.sck.mu.Lock()
+	rmw := srv.sck.w.(*routerMWriter)
+	for i := 0; i < 10; i++ {
+		w := &Conn{}
+		w.Peer.Meta = make(Metadata)
+		w.Peer.Meta[sysSockID] = id
+		w.rw = &sockSendEof{}
+		w.onCloseErrorCB = srv.sck.scheduleRmConn
+		// Do not to call srv.addConn as we dont want to have listener on this fake socket
+		rmw.addConn(w)
+		srv.sck.conns = append(srv.sck.conns, w)
+	}
+	srv.sck.mu.Unlock()
+
+	// Now try to send a message from the server to all clients.
+	msg := NewMsgFrom(nil, nil, []byte("payload"))
+	msg.Frames[0] = []byte(id)
+	if err := srv.Send(msg); err != nil {
+		t.Logf("Send to %s failed: %+v\n", id, err)
+	}
+}
+
+type sockSendEof struct {
+}
+
+func (r *sockSendEof) Write(b []byte) (n int, err error) {
+	runtime.Gosched()
```

Comment:
perhaps explain why we need this (and/or what it does)

Reply:
While was trying to explain figure out better simpler way here to trigger error
Now there is only two clients and no explicit call to scheduler.

You may ensure was deadlocking by commenting out actual patch at local version and run test.